controller/controller.go

```diff
@@ -237,9 +245,80 @@ func (c *Controller) deleteFunc(ctx context.Context, obj interface{}) {
 	}
 }
 
+// hashEvent takes an event and produces an integer hash that:
+// - is guaranteed to be stable for the lifetime of an object,
+// - is in the bounds 0 <= int < c.concurrency.
+func (c *Controller) hashEvent(ctx context.Context, event watch.Event) int {
+	v := reflect.ValueOf(event.Object).Elem()
+	objectMeta, ok := v.FieldByName("ObjectMeta").Interface().(metav1.ObjectMeta)
+	if !ok {
+		c.logger.LogCtx(ctx, "level", "error", "message", "could not get object meta from object", "object", fmt.Sprintf("%#v", event.Object))
+		return 0
+	}
+
+	key := fmt.Sprintf("%s/%s", objectMeta.GetNamespace(), objectMeta.GetName())
+
+	hash := fnv.New32a()
+	hash.Write([]byte(key))
+
+	ih := int(hash.Sum32())
+	h := ih % c.concurrency
+
+	return h
+}
+
 // ProcessEvents takes the event channels created by the operatorkit informer
-// and executes the controller's event functions accordingly.
+// and routes them to appropriate workers.
 func (c *Controller) ProcessEvents(ctx context.Context, deleteChan chan watch.Event, updateChan chan watch.Event, errChan chan error) error {
+	// eventChannels holds the channels necessary for sending events to a worker.
+	type eventChannels struct {
+		deleteChan chan watch.Event
+		updateChan chan watch.Event
+	}
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("starting %v workers", c.concurrency))
+
+	channels := make([]eventChannels, c.concurrency)
+
+	for i := 0; i < c.concurrency; i++ {
+		channels[i] = eventChannels{
+			deleteChan: make(chan watch.Event),
+			updateChan: make(chan watch.Event),
+		}
+
+		go c.processEvents(ctx, i, channels[i].deleteChan, channels[i].updateChan)
```

Comment:
I'm unsure of the context management here. When launching multiple goroutines, I'm worried that it's actually the same context being used everywhere.